src/lib.rs

```diff
@@ -371,246 +493,346 @@ struct Category<BlockNumber, Moment, AccountId> {
-
-            // Let's deposit an event to let the outside world know this happened.
-            Self::deposit_event(RawEvent::Dummy(increase_by));
+
+        /// Set forum sudo.
+        fn set_forum_sudo(newForumSudo: Option<T::AccountId>) -> dispatch::Result {
+
+            /*
+             * Question: when this routine is called by non sudo or with bad signature, what error is raised?
+             * Update ERror set in spec
+             */
+
+            // Hold on to old value
+            let oldForumSudo = <ForumSudo<T>>::get().clone();
+
+            // Update forum sudo
+            match newForumSudo.clone() {
+                Some(accountId) => <ForumSudo<T>>::put(accountId),
+                None => <ForumSudo<T>>::kill()
+            };
+
+            // Generate event
+            Self::deposit_event(RawEvent::ForumSudoSet(oldForumSudo, newForumSudo));
 
             // All good.
             Ok(())
         }
 
-        /// A privileged call; in this case it resets our dummy value to something new.
-        // Implementation of a privileged call. This doesn't have an `origin` parameter because
-        // it's not (directly) from an extrinsic, but rather the system as a whole has decided
-        // to execute it. Different runtimes have different reasons for allow privileged
-        // calls to be executed - we don't need to care why. Because it's privileged, we can
-        // assume it's a one-off operation and substantial processing/storage/memory can be used
-        // without worrying about gameability or attack scenarios.
-        // If you not specify `Result` explicitly as return value, it will be added automatically
-        // for you and `Ok(())` will be returned.
-        fn set_dummy(#[compact] new_value: BalanceOf<T>) {
-            // Put the new value into storage.
-            <Dummy<T>>::put(new_value);
-        }
+        /// Add a new category.
+        fn create_category(origin, parent: Option<CategoryId>, title: Vec<u8>, description: Vec<u8>) -> dispatch::Result {
 
-        // The signature could also look like: `fn on_initialize()`
-        fn on_initialize(_n: T::BlockNumber) {
-            // Anything that needs to be done at the start of the block.
-            // We don't do anything here.
-        }
+            // Check that its a valid signature
+            let who = ensure_signed(origin)?;
+
+            // Not signed by forum SUDO
+            Self::ensure_is_forum_sudo(&who)?;
+
+            // Validate title
+            ensure_category_title_is_valid(&title)?;
+
+            // Validate description
+            ensure_category_description_is_valid(&description)?;
 
-        // The signature could also look like: `fn on_finalize()`
-        fn on_finalize(_n: T::BlockNumber) {
-            // Anything that needs to be done at the end of the block.
-            // We just kill our dummy storage item.
-            <Dummy<T>>::kill();
+            if let Some(parent_category_id) = parent {
```

Comment:
Consider extracting the whole `if` block to a separate function called for example `verify_parent_category(...)`.
It will improve/simplify readability of `create_category` function.

Reply:
1. Its hard to to factor out any function for this scenario which is genuinely reusable, am I missing something? I currently does these three things in order
  a) check that parent exists
  b) check that parent or any ancestor is not deleted or archived 
  c) check that parent or any ancestor is not deleted or archived 
  - check that depth is not exceeded.

2. Or are you saying to move it out even if it's not reusable? 

3. I have given it a try, factored  out `ensure_can_add_subcategory_to_parent_at_path_tip`, also factored out `ensure_can_mutate_in_category` because it will need to be used for many other dispatchables.